controllers/provider-azure/pkg/azure/client/storage.go

```diff
@@ -19,17 +19,93 @@ import (
 	"fmt"
 	"net/url"
 
+	"github.com/Azure/azure-sdk-for-go/services/resources/mgmt/2019-05-01/resources"
+	"github.com/Azure/azure-sdk-for-go/services/storage/mgmt/2019-04-01/storage"
 	"github.com/Azure/azure-storage-blob-go/azblob"
+	"github.com/Azure/go-autorest/autorest/azure/auth"
 	"github.com/gardener/gardener-extensions/controllers/provider-azure/pkg/azure"
+	"github.com/gardener/gardener-extensions/controllers/provider-azure/pkg/internal"
 	extensionscontroller "github.com/gardener/gardener-extensions/pkg/controller"
 
 	corev1 "k8s.io/api/core/v1"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 )
 
+// NewStorageClientAuthFromSubsriptionSecretRef retrieves the azure storage client auth from specified by the secret reference.
+func NewStorageClientAuthFromSubsriptionSecretRef(ctx context.Context, c client.Client, secretRef *corev1.SecretReference, resourceGroupName, accountName, region string) (*StorageAuth, error) {
+	// Reference : https://github.com/Azure-Samples/azure-sdk-for-go-samples/blob/master/storage/account.go
+	clientAuth, err := internal.GetClientAuthData(ctx, c, *secretRef)
+	if err != nil {
+		return nil, err
+	}
+
+	groupsClient := resources.NewGroupsClient(clientAuth.SubscriptionID)
+	clientCredConfig := auth.NewClientCredentialsConfig(clientAuth.ClientID, clientAuth.ClientSecret, clientAuth.TenantID)
+	authorizer, err := clientCredConfig.Authorizer()
+	if err != nil {
+		return nil, err
+	}
+	groupsClient.Authorizer = authorizer
+	if _, err := groupsClient.CreateOrUpdate(ctx, resourceGroupName, resources.Group{
+		Location: &region,
+	}); err != nil {
+		return nil, err
+	}
+
+	storageAccountClient := storage.NewAccountsClient(clientAuth.SubscriptionID)
+	storageAccountClient.Authorizer = authorizer
+	future, err := storageAccountClient.Create(ctx, resourceGroupName, accountName, storage.AccountCreateParameters{
+		Sku: &storage.Sku{
+			Name: storage.StandardLRS,
+		},
+		Kind:     storage.BlobStorage,
+		Location: &region,
+		AccountPropertiesCreateParameters: &storage.AccountPropertiesCreateParameters{
+			AccessTier: storage.Cool,
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+	if err := future.WaitForCompletionRef(ctx, storageAccountClient.Client); err != nil {
+		return nil, err
+	}
+
+	keysResponse, err := storageAccountClient.ListKeys(ctx, resourceGroupName, accountName)
+	if err != nil {
+		return nil, err
+	}
+
+	key := (*keysResponse.Keys)[0]
```

Comment:
```suggestion
	key := keysResponse.Keys[0]
```

However, should be checked whether `keys != nil && len(keys) > 0`?

Reply:
There will be key associated with StorageAccount. So don't have to check that. The suggestion will lead to sysntex error. Its pointer to the array.

Reply:
Ah,ok, thanks. Alright!